File: semgus-sketch/src/main/java/org/semgus/sketch/ir/NonTerminal.java
package org.semgus.sketch.ir;

import org.jetbrains.annotations.NotNull;
import org.semgus.sketch.base.Param;

import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;

public record NonTerminal(
  @NotNull String name,
  @NotNull String relName,
  @NotNull List<Var> ordVars,
  @NotNull Map<String, List<Param>> params  // Attributes -> Parameters
) {

  public @NotNull String getNTVarName() {
    return ordVars.stream()
      .filter(v -> v.decl().type().name().equals(name))
      .findFirst()
      .orElseThrow(() -> new NoSuchElementException("Cannot find non-terminal variable name."))
      .varName();
  }
}

File: semgus-sketch/src/main/java/org/semgus/sketch/ir/NonTerminal.kt
package org.semgus.sketch.ir

import org.semgus.sketch.syntax.Param

/**
 * IR non-terminal.
 */
internal data class NonTerminal(
  val name: String,
  val relName: String,
  val ordVars: Sequence<Var>,
  val params: Map<String, Sequence<Param>>,
) {
  fun ntVarName() = ordVars
    .find { it.decl.type.name == name }
    ?.varName ?: throw NoSuchElementException("Cannot find non-terminal variable name.")

  fun inputs() = params["input"]
    ?: throw NoSuchElementException("Cannot find variables with the input attribute.")

  fun outputs() = params["output"]
    ?: throw NoSuchElementException("Cannot find variables with the output attribute.")
}
